Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+The mechanism by which an option line is chosen is unspecified.
+Only a single option line must be chosen.
+
+As the conditional controls the validity of the option for selection this information must be provided by implementing program to the parts of the game that makes the selections.
+The implementing program must not restrict the selection of invalid options.
+It is the responsibility of the other components of the game to control how invalid options are to be handled.
+
+#### Blocks
+
+Much like with flow control options may have blocks of statements which are triggered should that option line be chosen.
+Each option line may optionally have a block of statements associated with that option line.
+Similar again to the flow control, if an option line is selected its associated block of must be processed by the implementing program
+If an option isn't chosen the associated block must not be processed.
+
+Unlike the flow control however there is no clear way to tell apart different blocks and options from other parts of the yarn, instead indentation is used to determine blocks and the end of the options.
+The rules for this must be followed:
+
+The first option line in the options determines the base indentation for the options statement, this is determined by counting the number of whitespace elements before the `->` symbol.
+Any statements following the option line at a greater level of indentation counts as part of the block for that option line.
+Any other options lines with the same indentation is considered a new option line and closes the block for that option.
+
+These rules are repeated for each option line until a non-option line with the same, or less indentation than the base indentation is encountered which closes the block and the option statement entirely.
+
+Options can be nested inside options.
+Not every option line needs to have blocks.
+
+##### Tabs vs Space
+
+The choice to require either tabs or spaces over the other is unspecified.
+Tabs and spaces shouldn't be mixed.
+Should there be a need to convert between them the conversion rate must be the same at all points in the project.
+The rate of conversion between tabs to spaces, and spaces to tabs, is unspecified.
+If there is a need to choose one, tabs should be preferred due to their improved accessibility over spaces.
+
+## Expressions
+
+_Expressions_ are mathematical chains of variables, values, functions, expressions, and operators that produce a single value as output.
+
+Expressions are not a statement but are a component of various statements and must only be used as part of a statement, they cannot exist in isolation.
+This means if you do want to show the result of an expression it will have to be wrapped inside an interpolated line.
+For example a line that is just `$numberOfCoins + 1` is invalid, but `{$numberOfCoins + 1}` is valid.
+
+Expressions are mostly used to control the flow of the if statement.
```

Comment:
"Expressions are mostly used to control the flow of the if statement, though they can also be used as part of a set statement, and in interpolated dialogue."